README.md
yr.no-forecast
==============

![TravisCI](https://travis-ci.org/evanshortiss/yr.no-forecast.svg) [![npm version](https://badge.fury.io/js/yr.no-forecast.svg)](https://badge.fury.io/js/yr.no-forecast) [![Coverage Status](https://coveralls.io/repos/github/evanshortiss/yr.no-forecast/badge.svg?branch=master)](https://coveralls.io/github/evanshortiss/yr.no-forecast?branch=master)

Wrapper to easily get weather data for a specified location in JSON format. Uses [yr.no-interface](https://github.com/evanshortiss/yr.no-interface) under the
hood. See the API docs at [yr.no](http://api.yr.no/weatherapi/locationforecast/1.9/documentation).


## Usage
Use the ```getWeather(queryParams)``` function to get a
LocationForecast object by calling the "locationforecast" API.

Here's an example:


```js
const yrno = require('yr.no-forecast')({
  version: '1.9', // this is the default if not provided,
  request: {
    // make calls to locationforecast timeout after 15 seconds
    timeout: 15000
  }
});

const LOCATION = {
  // This is Dublin, Ireland
  lat: 53.3478,
  lon: 6.2597
};

yrno.getWeather(LOCATION)
  .then((weather) => {
    // Get general weather for next five days (Array with five objects)
    weather.getFiveDaySummary()
      .then((data) => console.log('five day summary', data));

    // Get a weather data point for a given time between now and 9 days ahead
    weather.getForecastForTime(new Date())
      .then((data) => console.log('current weather', data));
  })
  .catch((e) => {
    console.log('an error occurred!', e);
  });
```

## API

### module(config)
This module exports a single factory function that can be used to get a
configured `instance` that exports the `getWeather` function.

Currently supported config options:

* version - Passed when making a call to the met.no API to select the
locationforecast version to call
* request - Can be populated with options for the `request` module. The only
setting that you should need to pass is `timeout` and is demonstrated above


### instance.getWeather(params[, version])
Returns a Promise that will resolve with a `LocationForecast` object that
contains functions to get weather data. You can pass the version parameter if
you want to override the default of 1.9, or the default you supplied when
creating and instance.

### LocationForecast.getFiveDaySummary()
Returns a Promise that resolves to an Array of 5 weather data Objects.

### LocationForecast.getForecastForTime(time)
Returns a Promise that resolves to a weather data Object that is closest to the
provided `time` argument. The `time` argument will be passed to `moment.utc` so
many time formats will work, but a millisecond timestamp or ISO formatted date
string are both ideal options to use use.

### LocationForecast.getXml()
Returns the raw XML string that the `locationforecast` API returned.

### LocationForecast.getJson()
Returns the JSON representation of the entire `locationforecast` response.

### LocationForecast.getFirstDateInPayload()
Returns the first date string that is available in the data returned from the
locationforecast call.

### LocationForecast.getValidTimes()
Returns an Array of ISO timestamps that represent points in time that we have
weather data for.


## Weather JSON Format

Some fields will be undefined depending on the weather conditions. Always
verify the field you need exists, e.g use `data.hasOwnProperty('fog')` or
similar techniques.

```json
{
  "datatype": "forecast",
  "from": "2017-04-18T03:00:00Z",
  "to": "2017-04-18T03:00:00Z",
  "icon": "PartlyCloud",
  "rain": "0.0 mm",
  "altitude": "0",
  "latitude": "59.8940",
  "longitude": "10.6450",
  "temperature": {
    "id": "TTT",
    "unit": "celsius",
    "value": "-0.9"
  },
  "windDirection": {
    "id": "dd",
    "deg": "14.6",
    "name": "N"
  },
  "windSpeed": {
    "id": "ff",
    "mps": "1.5",
    "beaufort": "1",
    "name": "Flau vind"
  },
  "windGust": {
    "id": "ff_gust",
    "mps": "2.4"
  },
  "humidity": {
    "value": "78.3",
    "unit": "percent"
  },
  "pressure": {
    "id": "pr",
    "unit": "hPa",
    "value": "1030.1"
  },
  "cloudiness": {
    "id": "NN",
    "percent": "15.4"
  },
  "fog": {
    "id": "FOG",
    "percent": "0.0"
  },
  "lowClouds": {
    "id": "LOW",
    "percent": "15.4"
  },
  "mediumClouds": {
    "id": "MEDIUM",
    "percent": "0.8"
  },
  "highClouds": {
    "id": "HIGH",
    "percent": "0.0"
  },
  "dewpointTemperature": {
    "id": "TD",
    "unit": "celsius",
    "value": "-4.5"
  }
}
```

## CHANGELOG

Can be found [at this link](https://github.com/evanshortiss/yr.no-forecast/blob/master/CHANGELOG.md ).


CHANGELOG.md
# CHANGELOG

Dates use DD/MM/YYYY format.

## 2.1.0 - 12/04/2018
* Bump momentjs to avoid vulnerability
* Bump yr.no-interface to use latest api.met.no URL

## 2.0.0 - 24/05/2017
* Increase performance by approximately 10x due to use of `pixl-xml` and
improved algorithm for getting weather nodes.
* Simplify codebase.
* Update JSON output format to be closer to the original XML content. This gives
users of this module more flexibility and information but is a breaking change.
* Add `getValidTimes` function.
* Fix security `request` and `qs` module vulnerabilities by updating to `yr.no-interface@1.0.1`.

## 1.0.1 - 28/04/2017
* Mitigate security vulnerabilities.
* Use new yr.no-interface@1.0.0 internally.

## 1.0.0 - 27/04/2017
* Change to Promise based interface.
* Make the module a factory function.
* Use `pixl-xml` to increase XML parsing speed and lower memory usage. Yay for
better performance!
* Improved unit tests. Tests no longer require an internet connection.
* Removed `getCurrentSummary` in favour of simply using `getForecastForTime` and
passing the current timestamp. Simplifies API footprint.


## < 1.0.0
* Undocumented. Sorry.


test/test.js
'use strict';

const moment = require('moment');
const proxyquire = require('proxyquire');
const sinon = require('sinon');
const chai = require('chai');
const readFileSync = require('fs').readFileSync;
const join = require('path').join;

chai.use(require('chai-truthy'));
var expect = require('chai').expect;

const SAMPLE_XML = readFileSync(
  join(__dirname, '../fixtures/weather-response-oslo.xml'), 'utf8'
);


const LOCATION = {
  lat: 53.3478,
  lon: 6.2597
};

describe('yr.no-forecast', function() {
  this.timeout(5000);

  var lib, stubs, yrno;

  var YRNO = 'yr.no-interface';

  beforeEach(function () {
    yrno = {
      locationforecast: sinon.stub().yields(null, SAMPLE_XML)
    };

    stubs = {
      [YRNO]: sinon.stub().returns(yrno)
    };

    lib = proxyquire('../index.js', stubs);
  });

  describe('#getWeather', function () {
    it('should return an error for malformed xml payloads', function () {
      stubs[YRNO]({}).locationforecast.yields(null, '<xml><invalid </invalid></xml>');

      return lib().getWeather(LOCATION)
        .then(() => {
          throw new Error('not meant to be here');
        })
        .catch(function (err) {
          expect(err).to.be.truthy();
          expect(err.toString()).to.contain('Parse Error: Mismatched closing tag');

          expect(
            yrno.locationforecast.getCall(0).args[0]
          ).to.deep.equal({
            version: 1.9,
            query: LOCATION
          });
        });
    });

    it('should return an error on yr.no API errors', function () {
      stubs[YRNO]({}).locationforecast.yields(new Error('oh noes!'));

      return lib().getWeather(LOCATION)
        .then(() => {
          throw new Error('not meant to be here');
        })
        .catch(function (err) {
          expect(err).to.be.truthy();
          expect(err.toString()).to.contain('oh noes!');
        });
    });

    it('should return an object with LocationForecast methods', function() {
      return lib().getWeather(LOCATION)
        .then(function(weather) {
          expect(weather).to.be.an('object');
          expect(weather.xml).to.be.a('string');
          expect(weather.json).to.be.an('object');

          expect(weather.json.weatherdata).to.be.an('object');
          expect(weather.json.weatherdata.created).to.be.a('string');
          expect(weather.json.weatherdata.meta).to.be.an('object');
          expect(weather.json.weatherdata.product).to.be.an('object');
        });
    });

    it('should use custom version from config', () => {
      const version = 2.1;

      return lib({version: version}).getWeather(LOCATION)
        .then(function() {
          expect(
            yrno.locationforecast.getCall(0).args[0].version
          ).to.equal(version);
        });
    });

    it('should use custom version from args to getWeather', () => {
      const version = 2.1;

      return lib().getWeather(LOCATION, version)
        .then(function() {
          expect(
            yrno.locationforecast.getCall(0).args[0].version
          ).to.equal(version);
        });
    });
  });

  describe('#getXml', function () {
    it('should return a string', function () {
      return lib().getWeather(LOCATION)
        .then(function (weather) {
          expect(weather).to.be.truthy();
          expect(weather.getXml()).to.be.a('string');
        });
    });
  });


  describe('#getJson', function () {
    it('should return a value', function () {
      return lib().getWeather(LOCATION)
        .then(function (weather) {
          expect(weather).to.be.truthy();
          expect(weather.getJson()).to.be.an('object');
        });
    });
  });


  describe('#getFiveDaySummary', function () {
    it('should return array with 5 weather objects', function() {
      return lib().getWeather(LOCATION)
        .then(function(weather) {
          expect(weather).to.be.an('object');
          return weather.getFiveDaySummary();
        })
        .then(function (summary) {
          expect(summary).to.be.an('array');
          expect(summary).to.have.length(5);

          var prev = null;

          summary.forEach(function (cur) {
            if (prev) {
              expect(moment(prev.from).isBefore(moment(cur.from))).to.be.true;
            }

            prev = cur;

            expect(cur.icon).to.be.a('string');
            expect(cur.to).to.be.a('string');
            expect(cur.from).to.be.a('string');
            expect(cur.rain).to.be.a('string');
            expect(cur.temperature).to.be.an('object');
            expect(cur.windDirection).to.be.an('object');
            expect(cur.windSpeed).to.be.an('object');
            expect(cur.humidity).to.be.an('object');
            expect(cur.pressure).to.be.an('object');
            expect(cur.cloudiness).to.be.an('object');
            expect(cur.lowClouds).to.be.an('object');
            expect(cur.mediumClouds).to.be.an('object');
            expect(cur.highClouds).to.be.an('object');
            expect(cur.dewpointTemperature).to.be.an('object');
          });
        });
    });
  });

  describe('#getValidTimestamps', function () {
    it('should return an array of ISO timestamps', function () {
      return lib().getWeather(LOCATION)
        .then((weather) => weather.getValidTimestamps())
        .then(function (times) {
          expect(times).to.be.an('array');
          expect(times[0]).to.equal('2017-04-18T13:00:00Z');
        });
    });
  });

  describe('#getForecastForTime', function () {
    it('should return an error for invalid date arguments', function () {
      return lib().getWeather(LOCATION)
        .then(function(weather) {
          expect(weather).to.be.an('object');
          return weather.getForecastForTime('bad time string');
        })
        .catch(function (err) {
          expect(err).to.be.an('error');
          expect(err.toString()).to.contain(
            'Invalid date provided for weather lookup'
          );
        });
    });

    it('should return null if date is not in range', function () {
      return lib().getWeather(LOCATION)
        .then(function(weather) {
          expect(weather).to.be.an('object');

          return weather.getForecastForTime(moment.utc('2017-04-15'));
        })
        .then(function (forecast) {
          expect(forecast).to.equal(null);
        });
    });

    it('should get weather for 10:00 PM (22:00) by rounding up', function() {
      var time = moment.utc('2017-04-18');
      time.set('hours', 21);
      time.set('minutes', 35);

      return lib().getWeather(LOCATION)
        .then(function(weather) {
          return weather.getForecastForTime(time);
        })
        .then(function (forecast) {
          expect(forecast).to.be.an('object');
          expect(forecast.icon).to.be.a('string');
          expect(forecast.to).to.be.a('string');
          expect(forecast.from).to.be.a('string');
          expect(forecast.rain).to.be.a('string');
          expect(forecast.temperature).to.be.an('object');
          expect(forecast.windDirection).to.be.an('object');
          expect(forecast.windSpeed).to.be.an('object');
          expect(forecast.windGust).to.be.an('object');
          expect(forecast.humidity).to.be.an('object');
          expect(forecast.pressure).to.be.an('object');
          expect(forecast.cloudiness).to.be.an('object');
          expect(forecast.fog).to.be.an('object');
          expect(forecast.lowClouds).to.be.an('object');
          expect(forecast.mediumClouds).to.be.an('object');
          expect(forecast.highClouds).to.be.an('object');
          expect(forecast.dewpointTemperature).to.be.an('object');
        });
    });

    it('should get weather for 9:00 PM (21:00) by rounding down', function() {
      var time = moment.utc('2017-04-18');
      time.set('hours', 21);
      time.set('minutes', 20);

      return lib().getWeather(LOCATION)
        .then(function(weather) {
          return weather.getForecastForTime(time);
        })
        .then(function (forecast) {
          expect(forecast).to.be.an('object');
          expect(forecast.icon).to.be.a('string');
          expect(forecast.to).to.be.a('string');
          expect(forecast.from).to.be.a('string');
          expect(forecast.rain).to.be.a('string');
          expect(forecast.temperature).to.be.an('object');
          expect(forecast.windDirection).to.be.an('object');
          expect(forecast.windSpeed).to.be.an('object');
          expect(forecast.windGust).to.be.an('object');
          expect(forecast.humidity).to.be.an('object');
          expect(forecast.pressure).to.be.an('object');
          expect(forecast.cloudiness).to.be.an('object');
          expect(forecast.fog).to.be.an('object');
          expect(forecast.lowClouds).to.be.an('object');
          expect(forecast.mediumClouds).to.be.an('object');
          expect(forecast.highClouds).to.be.an('object');
          expect(forecast.dewpointTemperature).to.be.an('object');
        });
    });

    it('should get weather closest to 1:00 PM for 2017-04-25 using fallbackSelector', function () {
      var time = moment.utc('2017-04-25');
      time.set('hours', 13);
      time.set('minutes', 0);

      return lib().getWeather(LOCATION)
        .then(function(weather) {
          return weather.getForecastForTime(time);
        })
        .then(function (forecast) {
          expect(forecast).to.be.an('object');
          expect(forecast.from).to.equal('2017-04-25T12:00:00Z');
        });
    });
  });
});


example/dublin-weather.js
'use strict';

const yrno = require('../index.js')({
  version: '1.9', // this is the default if not provided,
  request: {
    // make calls to locationforecast timeout after 15 seconds
    timeout: 15000
  }
});

const LOCATION = {
  // This is Dublin, Ireland
  lat: 53.3478,
  lon: 6.2597
};

console.log('\nGetting weather for Dublin, Ireland...\n');

yrno.getWeather(LOCATION)
  .then((weather) => {
    // Get general weather for next five days (Array with five objects)
    // weather.getFiveDaySummary()
    //   .then((data) => console.log('\n five day summary', data));

    // Get a weather data point for a given time between now and 9 days ahead
    weather.getForecastForTime(new Date())
      .then((data) => {
        if (data.hasOwnProperty('temperature')) {
          console.log(`Temperature is around ${data.temperature}`);
        }

        if (data.hasOwnProperty('rain')) {
          console.log(`Expected rainfall is ${data.rain}`);
        }

        if (data.hasOwnProperty('humidity')) {
          console.log(`Humidity is ${data.humidity}`);
        }

        console.log('\n');
      });
  })
  .catch((e) => {
    console.log('an error occurred getting weather xml!', e);
  });


index.js
'use strict';

const log = require('debug')(require('./package.json').name);
const moment = require('moment');
const XML = require('pixl-xml');
const VError = require('verror');
const each = require('lodash.foreach');
const Promise = require('bluebird');

/**
 * "simple" nodes are those with very basic detail. 1 to 4 of these follow a
 * node with more details
 * @param  {Object}  node
 * @return {Boolean}
 */
function isSimpleNode (node) {
  return node.location.symbol !== undefined;
}

/**
 * Check if a node has a min and max temp range
 * @param  {Object}  node
 * @return {Boolean}
 */
function hasTemperatureRange (node) {
  return node.location.minTemperature && node.location.maxTemperature;
}

/**
 * Convert a momentjs date object into an ISO string compatible with our XML
 * @param  {Date}   date
 * @return {String}
 */
function dateToForecastISO (date) {
  return date.utc().format('YYYY-MM-DDTHH:mm:ss[Z]');
}

module.exports = (config) => {
  // Make a default config, but extend it with the passed config
  config = Object.assign({
    version: 1.9
  }, config);

  // Create a yrno instance with any overrides required
  const yrno = require('yr.no-interface')({
    request: config.request
  });

  return {
    /**
     * Public call to get a LocationForecast instance.
     * Version is optional.
     * @param {Object}    params
     * @param {String}    [version]
     */

    getWeather: (params, version) => {
      version = version || config.version;

      log('requesting a locationforecast using API version %s', version);

      return Promise.fromCallback(function (callback) {
        // Make a standard call to the API
        yrno.locationforecast({
          query: params,
          version: version
        }, function(err, body) {
          if (err) {
            log('failed to get locationforecast from yr.no API. Error:', err);
            return callback(err, null);
          }


          log('successfully retrieved locationforecast report from yr.no');

          // Wrap the response from API
          callback(null, new LocationForecast(body));
        });
      });
    }

  };
};

/**
 * @constructor
 * @param {String} xml
 */
function LocationForecast(xml) {
  this.xml = xml;

  // Map containing weather info for given utc times
  this.times = {
    // e.g '2017-04-29T01:00:00Z': { DATA HERE }
  };

  log('building LocationForecast object by parsing xml to JSON');

  var startDt = Date.now();

  // Parse to JSON and return this object on success
  try {
    this.json = XML.parse(xml, {preserveDocumentNode: true});
    log('parsing xml to json took %dms', Date.now() - startDt);
  } catch (e) {
    throw new VError(e, 'failed to parse returned xml string to JSON');
  }

  this._init();

  log('LocationForecast init complete in %sms', Date.now() - startDt);

  return this;
}

LocationForecast.prototype = {
  _init: function () {
    var self = this;

    each(this.json.weatherdata.product.time, function (node) {
      const simple = isSimpleNode(node);
      const temps = hasTemperatureRange(node);

      if (!simple) {
        self.times[node.to] = node;
      } else {
        // node is a small/simple node with format
        // <time datatype="forecast" from="2017-04-28T22:00:00Z" to="2017-04-28T23:00:00Z">
        //   <location altitude="17" latitude="34.0522" longitude="118.2437">
        //     <precipitation unit="mm" value="0.0"/>
        //     <symbol id="Sun" number="1"/>
        //   </location>
        // </time>

        const parent = self.times[node.to];

        parent.icon = node.location.symbol.id;
        parent.rain = node.location.precipitation.value + ' ' + node.location.precipitation.unit;

        parent.rainDetails = {
          minRain: 'minvalue' in node.location.precipitation ? node.location.precipitation.minvalue : null,
          maxRain: 'minvalue' in node.location.precipitation ? node.location.precipitation.maxvalue : null,
          rain: node.location.precipitation.value,
          unit: node.location.precipitation.unit
        };

        /* istanbul ignore else  */
        if (temps) {
          parent.minTemperature = node.location.minTemperature;
          parent.maxTemperature = node.location.maxTemperature;
        }
      }
    });
  },


  /**
   * Returns the JSON representation of the parsed XML`
   * @return {Object}
   */
  getJson: function() {
    return this.json;
  },


  /**
   * Return the XML string that the met.no api returned
   * @return {String}
   */
  getXml: function() {
    return this.xml;
  },


  /**
   * Returns the earliest ISO timestring available in the weather data
   * @return {String}
   */
  getFirstDateInPayload: function () {
    return this.json.weatherdata.product.time[0].from;
  },


  /**
   * Returns the latest ISO timestring available in the weather data
   * @return {String}
   */
  getLastDateInPayload: function () {
    return this.json.weatherdata.product.time[this.json.weatherdata.product.time.length - 1].from;
  },


  /**
   * Returns an array of all times that we have weather data for
   * @return {Array<String>}
   */
  getValidTimestamps: function () {
    return Object.keys(this.times);
  },


  /**
   * Get five day weather.
   * @param {Function} callback
   */
  getFiveDaySummary: function() {
    const startDate = moment.utc(this.getFirstDateInPayload());
    const baseDate = startDate.clone().set('hour', 12).startOf('hour');
    let firstDate = baseDate.clone();

    log(`five day summary is using ${baseDate.toString()} as a starting point`);

    /* istanbul ignore else  */
    if (firstDate.isBefore(startDate)) {
      // first date is unique since we may not have data back to midday so instead we
      // go with the earliest available
      firstDate = startDate.clone();
    }

    log(`getting five day summary starting with ${firstDate.toISOString()}`);

    return Promise.all([
      this.getForecastForTime(firstDate),
      this.getForecastForTime(baseDate.clone().add(1, 'days')),
      this.getForecastForTime(baseDate.clone().add(2, 'days')),
      this.getForecastForTime(baseDate.clone().add(3, 'days')),
      this.getForecastForTime(baseDate.clone().add(4, 'days'))
    ])
      .then(function (results) {
        // Return a single array of objects
        return Array.prototype.concat.apply([], results);
      });
  },


  /**
   * Verifies if the pased timestamp is a within range for the weather data
   * @param  {String|Number|Date}  time
   * @return {Boolean}
   */
  isInRange: function (time) {
    return moment.utc(time)
      .isBetween(
        moment(this.getFirstDateInPayload()),
        moment(this.getLastDateInPayload())
      );
  },


  /**
   * Returns a forecast for a given time.
   * @param {String|Date} time
   * @param {Function}    callback
   */
  getForecastForTime: function (time) {
    time = moment.utc(time);

    if (time.isValid() === false) {
      return Promise.reject(
        new Error('Invalid date provided for weather lookup')
      );
    }

    if (time.minute() > 30) {
      time.add('hours', 1).startOf('hour');
    } else {
      time.startOf('hour');
    }

    log('getForecastForTime', dateToForecastISO(time));

    let data = this.times[dateToForecastISO(time)] || null;

    /* istanbul ignore else  */
    if (!data && this.isInRange(time)) {
      data = this.fallbackSelector(time);
    }

    /* istanbul ignore else  */
    if (data) {
      data = Object.assign({}, data, data.location);
      delete data.location;
    }

    return Promise.resolve(data);
  },


  fallbackSelector: function (date) {
    log('using fallbackSelector for date', date);

    const datetimes = Object.keys(this.times);

    let closest = null;
    let curnode, curTo;
    let len = datetimes.length - 1;

    while (len) {
      curnode = this.times[datetimes[len]];
      curTo = moment(curnode.to);

      if (date.isSame(curTo, 'day')) {
        if (!closest) {
          closest = curnode;
        } else {
          /* istanbul ignore else  */
          if (Math.abs(date.diff(curTo)) < Math.abs(date.diff(moment(closest.to)))) {
            closest = curnode;
          }
        }
      } else if (closest) {
        // we found a node, and no more nodes exist for the day we need...BAIL
        break;
      }

      len--;
    }

    return closest;
  }
};
